airflow/providers/microsoft/azure/operators/azure_data_factory.py

```diff
@@ -0,0 +1,175 @@
+        self.start_activity_name = start_activity_name
+        self.start_from_failure = start_from_failure
+        self.parameters = parameters
+        self.wait_for_completion = wait_for_completion
+        self.timeout = timeout
+        self.poke_interval = poke_interval
+
+    def execute(self, context: Dict) -> None:
+        self.hook = AzureDataFactoryHook(conn_id=self.conn_id)
+        self.log.info(f"Executing the '{self.pipeline_name}' pipeline.")
+        response = self.hook.run_pipeline(
+            pipeline_name=self.pipeline_name,
+            resource_group_name=self.resource_group_name,
+            factory_name=self.factory_name,
+            reference_pipeline_run_id=self.reference_pipeline_run_id,
+            is_recovery=self.is_recovery,
+            start_activity_name=self.start_activity_name,
+            start_from_failure=self.start_from_failure,
+            parameters=self.parameters,
+        )
+        self.run_id = vars(response)["run_id"]
+        # Push the ``run_id`` value to XCom regardless of what happens during execution. This allows for
+        # retrieval the executed pipeline's ``run_id`` for downstream tasks especially if performing an
+        # asynchronous wait.
+        context["ti"].xcom_push(key="run_id", value=self.run_id)
+
+        if self.wait_for_completion:
+            self.log.info(f"Waiting for run ID {self.run_id} of pipeline {self.pipeline_name} to complete.")
+            start_time = time.monotonic()
+            pipeline_run_status = None
+            while pipeline_run_status not in AzureDataFactoryPipelineRunStatus.TERMINAL_STATUSES:
+                # Check to see if the pipeline-run duration has exceeded the ``timeout`` configured.
+                if start_time + self.timeout < time.monotonic():
+                    raise AirflowException(
+                        f"Pipeline run {self.run_id} has not reached a terminal status after "
+                        f"{self.timeout} seconds."
+                    )
+
+                # Wait to check the status of the pipeline based on the ``poke_interval`` configured.
+                time.sleep(self.poke_interval)
+
+                self.log.info(f"Checking on the status of run ID {self.run_id}.")
+                pipeline_run = self.hook.get_pipeline_run(
+                    run_id=self.run_id,
+                    factory_name=self.factory_name,
+                    resource_group_name=self.resource_group_name,
+                )
+                pipeline_run_status = pipeline_run.status
+                self.log.info(f"Run ID {self.run_id} is in a status of {pipeline_run_status}.")
+
+            if pipeline_run_status == AzureDataFactoryPipelineRunStatus.CANCELLED:
+                raise AirflowException(f"Pipeline run {self.run_id} has been cancelled.")
+
+            if pipeline_run_status == AzureDataFactoryPipelineRunStatus.FAILED:
+                raise AirflowException(f"Pipeline run {self.run_id} has failed.")
```

Comment:
This feels like a lot of logic that just shouldn't been in operator level.
I would suggest to take a look at Tableau provider. It's almost identical case.

https://github.com/apache/airflow/blob/95279bdb8989fc0e433799792dab2ed38373e1f8/airflow/providers/tableau/operators/tableau.py#L123-L129

Reply:
Fair point. I'll refactor this piece of the operator and sensor.